pkg/kubelet/volumemanager/volume_manager.go

```diff
@@ -334,6 +335,31 @@ func (vm *volumeManager) WaitForAttachAndMount(pod *v1.Pod) error {
 		return nil
 	}
 
+	volumeModeMap, err := makeVolumeModeMap(pod, vm.kubeClient)
```

Comment:
This new code checks that pod's VolumeDevices refers to Block PVCs. It's correct, but why it's in `WaitForAttachAndMount`? And should it check also `VolumeMounts` too?

IMO this check should be somewhere else. There are admission hooks in kubelet that allow a pod to run on the node or they refuse to run the pod with an error message to the user. Maybe a new handler that would check modes of all PVCs and VolumeMounts and VolumeDevices would be more appropriate. Someone from sig-node should confirm that PVCs are bound at this point (IMO they are...)

Reply:
@jsafrane 
Thank you for the suggestion. Let me investigate the admission hooks.
The purpose of this check is invalid parameter combinations should be return as error before as early as possible.(At least, we should return error before creating a volume)
If we can check this at API validation phase, it's nice but seems we can't check this at validation phase.

>> And should it check also VolumeMounts too?

Agree.

Reply:
In the proposal PR, the discussion concluded that we can't verify at API validation or admission controller.  So we have to do it early in kubelet.

Reply:
@msau42 thanks, so we should check this in the WaitForAttachAndMount(), right?

Reply:
So the check for if PVC is bound is done in [here](https://github.com/kubernetes/kubernetes/blob/50ec4384ff0ac482beb164767e562765ad82bde7/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L339).  We don't add the volume to dsw PVC is not bound.

We could potentially extend this check to verify the `volumeMode`.  I would also generate an event on the Pod with the error.  Otherwise they have to wait for the attach/mount failure with a generic "timed out" message.

Reply:
Kubelet has its own admission, it's something else than admission controller on the API server.  Kubelet admission runs after a pod gets scheduled on a node and kubelet can say "no, I won't run this" from whatever reason. The pod is then Pending forever with appropriate events / messages. 

Look for `PodAdmitHandler` interface in kubelet and all `AddPodAdmitHandler` calls out there.

Note that I just know that this mechanism exists, I don't know if it's the right tool to use here. Ask someone from sig-node.

Reply:
Pod rejection will send it back to the scheduler and will cause it to endlessly loop between scheduler and kubelet.